Add `pb_read` and `pb_write` functions

tanho63 opened the pull request:
Closes #97. 

I thought briefly about making this a wrapper around pb_download_url + read function that accepts URLs, but I don't think it had the flexibility I wanted plus I ran into issues downloading from private repositories that I later learned was around not being able to pass an auth token to it. 

I think this is the most flexible approach to the problem but would love to hear any thoughts

tanho63 committed:
add pb_read/pb_write and tests

tanho63 committed:
bump news.md

tanho63 committed:
typos

cboettig reviewed (changes requested):
this strategy (specifically the `guess_read_function()`) always makes me nervous.  While all abstractions are a bit leaky, I've always found attempts to write a generic "read" function as an abstract method particularly leaky.  There's just too much going on in `read_` methods to abstract away (what about other data serializations, like spatial formats? what about lazy reads / remote reads etc).  

I'm fine with a convenience wrapper that makes common use patterns more concise.  Maybe all that's needed is a bit more documentation saying something like 'for common formats such as ...`, along with advice about how to bypass this convenience if a user prefers their own read function (say, from `readr` package, or for some other format).

tanho63 commented:
> too much going on in read_ methods to abstract away (what about other data serializations, like spatial formats? 

I believe this will fail on spatial formats since it would not be one of csv, tsv, rds, parquet? (unfamiliar with how geoparquet works and whether arrow::read_parquet will process geoparquet, but I assume yes?). 

> what about lazy reads / remote reads etc?

Yep, this will read eagerly by design/default, and maybe it's a bad thing for folks who should be thinking about optimizing - however uninformed users would currently do pb_download anyways so it's not necessarily much different than that? 

I agree with improving the docs, e.g. 
- explaining that cloud native is best performance (more clearly linking to vignette)
- demonstrating ways to read from URL (maybe adding to getting started vignette?)
- better demonstrating examples of passing in a different read function and explaining what is supported ?

tanho63 committed:
bugfix example for pb_write

tanho63 committed:
stash work on vignette

tanho63 committed:
improve pb_read and pb_write documentation

tanho63 commented on R/pb_read.R:
```diff
@@ -0,0 +1,90 @@
+#' @param repo string: GH repository name in format "owner/repo". Default
+#' `guess_repo()` tries to guess based on current working directory's git repo
+#' @param tag  string: tag for the GH release, defaults to "latest"
+#' @param read_function function: used to read in the data, where the file is
+#' passed as the first argument and any additional arguments are subsequently
+#' passed in via `...`. Default `guess_read_function(file)` will check the file
+#' extension and try to find an appropriate read function if the extension is one
+#' of rds, csv, tsv, parquet, txt, or json, and will abort if not found.
+#' @param ... additional arguments passed to `read_function` after file
+#' @param .token GitHub authentication token, see [gh::gh_token()]
```
Fleshed out more of this section, explaining how the read_function will be used (file as first arg etc)

tanho63 commented on R/pb_read.R:
```diff
@@ -0,0 +1,90 @@
+
+#' Guess read function from file extension
+#'
+#' This function accepts a filename and tries to return a valid function for
+#' reading it.
+#'
+#' `guess_read_function` understands the following file extensions:
+#' - rds with `readRDS`
+#' - csv, csv.gz, csv.xz with `utils::read.csv`
+#' - tsv, tsv.gz, tsv.xz with `utils::read.delim`
+#' - parquet with `arrow::read_parquet`
+#' - txt, txt.gz, txt.xz with `readLines`
+#' - json, json.gz, json.xz with `jsonlite::fromJSON`
+#'
+#' @family pb_rw
+#' @param file filename to parse
+#' @return function for reading the file, if found
+#' @keywords internal
```
`guess_read_function` also gets a man page with explicit mappings

tanho63 commented on R/pb_write.R:
```diff
@@ -0,0 +1,84 @@
+    file_ext,
+    "rds" = saveRDS,
+    "csv" = utils::write.csv,
+    "tsv" = function(x, file, ..., sep = "\t") utils::write.csv(x = x, file = file, sep = sep, ...),
```
I don't use tsvs a lot, is there a better tsv writing function I don't know about?

tanho63 commented on vignettes/piggyback.Rmd:
```diff
@@ -145,6 +147,10 @@ pb_download_url(repo = "cboettig/piggyback-tests",
+Note that these will not work if authentication is required to download them
+
+## Reading data
+
```
in progress, will finish later this weekend hopefully. 

Will try to add examples of reading URLs above, and then will demonstrate pb_read/write below including how to pass it other read_functions.

tanho63 committed:
update pkgdown and ignore future local pkgdown builds

tanho63 committed:
remove git-tracking of pkgdown since ropensci builds it separately

tanho63 committed:
switch to standard rmarkdown chunks with eval=FALSE

tanho63 committed:
shorten DESCRIPTION description based on new readme

tanho63 committed:
reknit readme

tanho63 committed:
start adding section on reading in URLs

tanho63 committed:
relocate dots to before named args

tanho63 commented on vignettes/piggyback.Rmd:
````diff
@@ -36,113 +33,202 @@ number of files or bandwidth to deliver them.
-[2] "https://github.com/cboettig/piggyback-tests/releases/download/v0.0.1/iris.tsv.gz"    
-[3] "https://github.com/cboettig/piggyback-tests/releases/download/v0.0.1/iris.tsv.xz"    
-[4] "https://github.com/cboettig/piggyback-tests/releases/download/v0.0.1/iris2.tsv.gz" 
+Note that using `url()` requires that we close the connection after reading it, 
+or else we will receive warnings about leaving open connections.
+
+We can also use this `url()` approach to pass along authentication for private 
+repositories, e.g.
+```{r}
+pb_url <- pb_download_url("mtcars.rds", repo = "tanho63/piggyback-private", url_type = "api") %>%
+  url(
+    headers = c(
+      "Accept" = "application/octet-stream",
+      "Authorization" = paste("Bearer", gh::gh_token())
+    )
+  )
+readRDS(pb_url)
+#> # A tibble: 32 × 11
+#>     mpg   cyl  disp    hp  drat    wt  qsec    vs    am  gear  carb
+#>   <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
+#> 1  21       6   160   110  3.9   2.62  16.5     0     1     4     4
+#> 2  21       6   160   110  3.9   2.88  17.0     0     1     4     4
+#> 3  22.8     4   108    93  3.85  2.32  18.6     1     1     4     1
+#> 4  21.4     6   258   110  3.08  3.22  19.4     1     0     3     1
+#> 5  18.7     8   360   175  3.15  3.44  17.0     0     0     3     2
+#> # ℹ 27 more rows
+#> # ℹ Use `print(n = ...)` to see more rows
+close(pb_url)
````
@cboettig started writing out this "how to read files from URL" section and I feel like this got really complicated really quickly (esp with private repo URLs). This makes me think we should have a pb_read_url that handles both public and private repo URLs, but, unlike with just downloading it to disk, URL support is more complicated especially if we want to abstract the authentication for them:

- base R stuff tends to accept plain URLs and `url()` except `readRDS` which only works with `url()`
- jsonlite supports a url() connection or plain URL
- arrow doesn't accept a `url()` connection but instead wants the plain URL or a raw vector, so we would need to curl it first to pass the auth headers

```r
guess_read_url_function <- function(file) {
  switch(
    file_extension,
    "csv" = function(url, ..., headers) {con<-url(url, headers = headers); on.exit(close(con)); read.csv(con,...)},
    "rds" = function(url, ..., headers) {con<-url(url, headers = headers); on.exit(close(con)); readRDS(con,...)},
    "parquet" = function(url, ..., headers) {httr::GET(url, headers = headers)$content %>% read_parquet(...)},
    stop("unknown fmt")
)
```

tanho63 replied:
two paths-ish forward: 

A) tell users to just go to `pb_read` for simplicity if the repo is private
B) make `pb_read_url` happen

I'm ambivalent between the two and unsure how much difference in performance there actually is between pb_read and hypothetical pb_read_url paths...actually I can test that:

tanho63 replied:
alright this surprises the heck out of me:

```r
bench::mark(
  pb_read = pb_read("mtcars.csv", repo = "tanho63/piggyback-tests"),
  read.csv = read.csv(pb_download_url("mtcars.csv", repo = "tanho63/piggyback-tests")),
  read_csv = readr::read_csv(pb_download_url("mtcars.csv", repo = "tanho63/piggyback-tests")),
  fread = data.table::fread(pb_download_url("mtcars.csv", repo = "tanho63/piggyback-tests")),
  arrow = arrow::read_csv_arrow(pb_download_url("mtcars.csv", repo = "tanho63/piggyback-tests")),
  check = FALSE,
  iterations = 5
)
# A tibble: 5 × 13                                                                                   
#> expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time 
#> <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> 
#>   1 pb_read       193ms    206ms      4.90  265.87KB     0        5     0      1.02s 
#>   2 read.csv      603ms    618ms      1.57  147.84KB     0        5     0      3.18s 
#>   3 read_csv      427ms    454ms      2.25    2.09MB     1.50     3     2      1.34s 
#>   4 fread         140ms    143ms      6.50   33.86KB     0        5     0   769.03ms 
#>   5 arrow         604ms    620ms      1.57    2.02MB     0        5     0      3.18s 
```

tanho63 replied:
Attempting with larger size...

```r
options(piggyback.verbose = FALSE)
bench::mark(
  pb_read = pb_read("play_by_play_2023.csv", repo = "nflverse/nflverse-data", tag = "pbp"),
  read.csv = read.csv(pb_download_url("play_by_play_2023.csv", repo = "nflverse/nflverse-data", tag = "pbp")),
  read_csv = readr::read_csv(pb_download_url("play_by_play_2023.csv", repo = "nflverse/nflverse-data", tag = "pbp")),
  fread = data.table::fread(pb_download_url("play_by_play_2023.csv", repo = "nflverse/nflverse-data", tag = "pbp")),
  check = FALSE,
  min_iterations = 2
)
#> # A tibble: 4 × 13                                                               
#> expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
#> <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
#> 1 pb_read       7.51s    7.62s     0.131     583MB    0.787     2    12     15.24s
#> 2 read.csv      5.52s    5.97s     0.168     582MB    0.838     2    10     11.93s
#> 3 read_csv      4.54s    5.64s     0.177     200MB    0.266     2     3     11.28s
#> 4 fread         3.99s    4.13s     0.242     215MB    0.484     2     4      8.27s
```
okay, phew, must have been overhead or something

tanho63 replied:
Opted for option 1 so that I don't have to implement pb_read_url right away. This PR got away from me again...

tanho63 commented on vignettes/piggyback.Rmd:
````diff
@@ -36,113 +33,202 @@ number of files or bandwidth to deliver them.
+#> ├── iris.tsv.xz
+#> └── iris2.tsv.gz
 ```
+5. The `use_timestamps` argument defaults to TRUE - notice that above, 
````
noticed while explaining this that use_timestamps and overwrite are not merged as they are in pb_upload, raised new issue in #119

tanho63 committed:
try to align diffs

tanho63 committed:
add pb_read section

tanho63 committed:
add pb_write and edits to pb_upload

tanho63 commented:
Flow state hit me like a bus, many apologies for this PR running away from me. Since your last review ([diff](https://github.com/ropensci/piggyback/pull/115/files/6ff7768a3a148d6b6c7dfc58821bd193a8cfa08b..8f8d08d622f5655dedd90c49d8d23f42cee894c1)), I:
- gitignored + got rid of the stored `docs/` folder because ropensci builds pkgdown externally + I've been using it to test out how the vignette looks after generating it
- updated DESCRIPTION with the newly shortened description from the readme (because I realized it hadn't been updated)
- fleshed out man files for read_function and write_function args, added man files for guess_read_function and guess_write_function that fully explain how it gets mapped
- added various \dontshow blocks to hide the interactive() and try() blocks as inspired by the examplesIf 
- updated README to not evaluate any chunks except for regenerating codemeta (improves syntax highlighting in RStudio to use {r} instead of just r)
- rewrote vignette/piggyback.Rmd again to try and address your various feedback points - this is probably the [main thing](https://github.com/ropensci/piggyback/pull/115/files/6ff7768a3a148d6b6c7dfc58821bd193a8cfa08b..8f8d08d622f5655dedd90c49d8d23f42cee894c1#diff-795726fefb581909ff7dbfcc9a68590b225463c8b0836491f99652ac6c810981) that needs looked at

cboettig reviewed (approved):
looks good!
